README.md
# cleo


CHANGELOG.md
# Changelog for `cleo`

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to the
[Haskell Package Versioning Policy](https://pvp.haskell.org/).

## Unreleased

## 0.1.0.0 - YYYY-MM-DD


src/Lib.hs
module Lib
  ( repl
  ) where

import System.IO (hFlush, stdout)
import Control.Monad (unless)
import Control.Exception (try, SomeException, IOException, evaluate)
import Data.Char (isAlphaNum, isSpace)
import Data.Typeable (typeOf)
import System.Process (callCommand, readProcessWithExitCode, shell, createProcess, readCreateProcessWithExitCode)
import System.Exit (ExitCode (..))

import Functions.Time
import Functions.Weather
import Functions.Utility
import Functions.Joke

prompt :: String
prompt = ">> "

exit :: String
exit = ":q"

introduction :: String
introduction = "This is Cleo, your terminal assistant.\
                \ Cleo learns from your terminal\
                \ commands and starts suggesting\
                \ commands as you need them. When in doubt,\
                \ ask Cleo! Type ':q' anytime to quit."

defaultResponse = [ "Sorry, I don't know how to respond to that."
                  , "Sorry, I don't understand you."
                  , "I don't know how to do that." ]

read' :: IO String
read' = do
  putStr prompt
  hFlush stdout
  getLine

preprocess :: String -> [String]
preprocess input = map (filter isAlphaNum) . words $ input

eval' :: String -> IO ()
{-
eval' input = do
  output <- try (readProcessWithExitCode input [] []) :: IO (Either SomeException (ExitCode, String, String))
  case output of
    Right (ExitSuccess, stdout, _) -> putStrLn stdout
    Right (ExitFailure e, _, stderr) -> putStrLn $ "Process Exited with Exit Code " ++ show e ++ ": " ++ stderr
    Left e -> putStrLn "Error"
-}
eval' input = do
  (exitcode, stdout, stderr) <- readCreateProcessWithExitCode (shell input) ""
  case exitcode of
    ExitSuccess -> putStrLn stdout
    ExitFailure e
      | e == 127 -> matchResponse input
      | otherwise -> putStrLn $ stderr ++ "\nExited with Exit Code: " ++ show e

matchResponse :: String -> IO ()
matchResponse input 
  | "time" `elem` tokens = time
  | "weather" `elem` tokens = weather
  | "joke" `elem` tokens = joke
  | otherwise = putStrLn "Sorry, I don't understand."
  where tokens = preprocess input

{-
eval' input = do
  output <- try . callCommand $ input :: IO (Either IOException ())
  case output of
    Right success -> putStrLn "Success"
    Left e
      | "time" `elem` tokens -> time
      | "weather" `elem` tokens -> weather
      | otherwise -> putStrLn "Sorry, I don't understand."
      where tokens = preprocess input
-}
{-
eval' input
  | "time" `elem` tokens = time
  | "weather" `elem` tokens = weather
  | otherwise = do
      output <- try . callCommand $ input :: IO (Either SomeException ())
      case output of
        Left e -> putStrLn "Exception raised."
        Right success -> return success
  where
    tokens = preprocess input
-}

repl :: IO ()
repl = do
  putStrLn introduction
  repl'

repl' :: IO ()
repl' = do
  input <- read'
  unless (input == exit) $ do
    eval' input
    repl'


src/Functions/Joke.hs
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE NamedFieldPuns #-}

module Functions.Joke (joke) where

import qualified Data.Text as T
import Network.Wreq
import Data.Aeson
import Control.Lens

joke = undefined

weather :: IO ()
weather = do
    config' <- readFile ".weatherConfig"
    let config = getConfig config'
    getWeather config

data Config = Config { apiKey    :: T.Text
                     , latitude  :: T.Text
                     , longitude :: T.Text }

data WeatherResponse = WeatherResponse { main        :: T.Text
                                       , description :: T.Text
                                       , temperature :: Double
                                       , feelsLike   :: Double
                                       , pressure    :: Double
                                       , humidity    :: Double
                                       , visibility  :: Double
                                       , cloudiness  :: Double
                                       , name        :: T.Text }

instance Show WeatherResponse where
    show response = string
      where
        string = T.unpack . T.concat $ [ name $ response
                                       , "\n"
                                       , T.toTitle . description $ response
                                       , " at "
                                       , joinDouble temperature $ "°C"
                                       , ", feels like "
                                       , joinDouble feelsLike $ "°C"
                                       , ".\nPressure at "
                                       , joinDouble pressure $ "hPa"
                                       , " and humidity at "
                                       , joinDouble humidity $ "%"
                                       , ".\nVisibility is "
                                       , joinDouble visibility $ "m"
                                       , " and cloudiness is "
                                       , joinDouble cloudiness $ "%"
                                       , "." ]
        bold text = T.concat ["\\033[1m", text, "\\033[0m"]
        joinDouble f unit = T.pack $ show (f  response) ++ unit
    
instance FromJSON WeatherResponse where
    parseJSON = withObject "weather response" f
      where
        f o = do
            weatherO     <- o               .: "weather"
            main         <- (head weatherO) .: "main"
            description  <- (head weatherO) .: "description"
            temperatureO <- o               .: "main"
            temperature  <- temperatureO    .: "temp"
            feelsLike    <- temperatureO    .: "feels_like"
            pressure     <- temperatureO    .: "pressure"
            humidity     <- temperatureO    .: "humidity"
            visibility   <- o               .: "visibility"
            cloudsO      <- o               .: "clouds"
            cloudiness   <- cloudsO         .: "all"
            name         <- o               .: "name"
            return WeatherResponse { main
                                   , description
                                   , temperature
                                   , feelsLike
                                   , pressure
                                   , humidity
                                   , visibility
                                   , cloudiness
                                   , name }

getConfig :: String -> Config
getConfig file = Config { apiKey
                        , latitude
                        , longitude }
  where
    splitStrip                    = (map T.strip) . T.splitOn "="
    pairs                         = map splitStrip . T.lines . T.pack $ file
    [apiKey, latitude, longitude] = map (!! 1) pairs

getWeather :: Config -> IO ()
getWeather config = do
    r <- ((get . T.unpack) (endpoint))
    case decode (r ^. responseBody) of
        Just w  -> print (w :: WeatherResponse)
        Nothing -> fail "Weather JSON could not be parsed"
  where
    endpoint = T.concat
        [ "https://api.openweathermap.org/data/2.5/weather?lat="
        , latitude config
        , "&lon="
        , longitude config
        , "&appid="
        , apiKey config
        , "&units=metric" ]


src/Functions/Weather.hs
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE NamedFieldPuns #-}

module Functions.Weather (weather) where

import qualified Data.Text as T
import Network.Wreq
import Data.Aeson
import Control.Lens
import Control.Exception (IOException, try)

filePath = ".weatherConfig"

weather :: IO ()
weather = do
    config' <- try . readFile $ ".weatherConfig" :: IO (Either IOException String)
    case config' of
      Left _ -> putStrLn $ "Error: Config file path '" ++ filePath ++ "' could not be found.\n"
      Right config' -> do
        let config = getConfig config'
        getWeather config

data Config = Config { apiKey    :: T.Text
                     , latitude  :: T.Text
                     , longitude :: T.Text }

data WeatherResponse = WeatherResponse { main        :: T.Text
                                       , description :: T.Text
                                       , temperature :: Double
                                       , feelsLike   :: Double
                                       , pressure    :: Double
                                       , humidity    :: Double
                                       , visibility  :: Double
                                       , cloudiness  :: Double
                                       , name        :: T.Text }

instance Show WeatherResponse where
    show response = string
      where
        string = T.unpack . T.concat $ [ name $ response
                                       , "\n"
                                       , T.toTitle . description $ response
                                       , " at "
                                       , joinDouble temperature "°C"
                                       , ", feels like "
                                       , joinDouble feelsLike "°C"
                                       , ".\nPressure at "
                                       , joinDouble pressure "hPa"
                                       , " and humidity at "
                                       , joinDouble humidity "%"
                                       , ".\nVisibility is "
                                       , joinDouble visibility "m"
                                       , " and cloudiness is "
                                       , joinDouble cloudiness "%"
                                       , "." ]
        bold text = T.concat ["\\033[1m", text, "\\033[0m"]
        joinDouble f unit = T.pack $ show (f  response) ++ unit
    
instance FromJSON WeatherResponse where
    parseJSON = withObject "weather response" f
      where
        f o = do
            weatherO     <- o               .: "weather"
            main         <- (head weatherO) .: "main"
            description  <- (head weatherO) .: "description"
            temperatureO <- o               .: "main"
            temperature  <- temperatureO    .: "temp"
            feelsLike    <- temperatureO    .: "feels_like"
            pressure     <- temperatureO    .: "pressure"
            humidity     <- temperatureO    .: "humidity"
            visibility   <- o               .: "visibility"
            cloudsO      <- o               .: "clouds"
            cloudiness   <- cloudsO         .: "all"
            name         <- o               .: "name"
            return WeatherResponse { main
                                   , description
                                   , temperature
                                   , feelsLike
                                   , pressure
                                   , humidity
                                   , visibility
                                   , cloudiness
                                   , name }

getConfig :: String -> Config
getConfig file = Config { apiKey
                        , latitude
                        , longitude }
  where
    splitStrip                    = (map T.strip) . T.splitOn "="
    pairs                         = map splitStrip . T.lines . T.pack $ file
    [apiKey, latitude, longitude] = map (!! 1) pairs

getWeather :: Config -> IO ()
getWeather config = do
    r <- ((get . T.unpack) (endpoint))
    case decode (r ^. responseBody) of
        Just w  -> print (w :: WeatherResponse)
        Nothing -> fail "Weather JSON could not be parsed"
  where
    endpoint = T.concat
        [ "https://api.openweathermap.org/data/2.5/weather?lat="
        , latitude config
        , "&lon="
        , longitude config
        , "&appid="
        , apiKey config
        , "&units=metric" ]
